### DISCUSSION TITLE
Unable to see structured logs in pubished app

### DISCUSSION THREAD

**Discussion by efonsecab** [microsoft/aspire#4583] [Q&A]
I'm not being able to see the structured logs in the published app, they however show correctly locally.

![image](https://github.com/dotnet/aspire/assets/3481899/b3a7d964-7698-4aac-b73a-a69a015f4c57)

Same is happening for traces and metrics.

This is the dashboard version

![image](https://github.com/dotnet/aspire/assets/3481899/41bcb284-3791-4b1b-b259-34017ee87135)

AppHost project
![image](https://github.com/dotnet/aspire/assets/3481899/9f06f2af-40ae-4963-a816-9051506f8468)

ServiceDefaults project
![image](https://github.com/dotnet/aspire/assets/3481899/26e6988b-9b13-42a1-8e2f-9564fecb5bfa)

Log Streams are not showing any error, so no idea what may be causing it.

**Comment by mitchdenny** [comment]
If you publishing the `aspire-starter` template does it work for you? Just trying to narrow down whether we have a fundamental bug here or it is just something specific to your code.

**Comment by mitchdenny** [comment]
@efonsecab I am moving this issue over to discussions because I would be surprised if this is a bug given how often we hit this code path. If after some discussion it does end up being a bug we'll open an issue with the specifics of the bug.

**Reply by efonsecab** [reply]
> @efonsecab I am moving this issue over to discussions because I would be surprised if this is a bug given how often we hit this code path. If after some discussion it does end up being a bug we'll open an issue with the specifics of the bug.

I think it has to do with my code since I recently upgraded some packages, though I still make sure to keep Aspire in 8.0.1, the main issue however is that the log streams are not showing any errors, and as far as I know the dashboard does not have information logs for itself, so I have no idea how to debug the issue other than guessing packages versions combinations.

**Comment by MrDeej** [comment]
Same problem for my. Logging/tracing/metrics works in localhost and my test environment. But not in my production envionment. And I dont find anything that explains why. I have no enviroment specific coding outside of the starter kit.

**Comment by kevcrooks** [comment]
This has started happening for me too. If I start a new `Aspire` project, the structured logs appear, but I have a project that has already been published and the structured logs have stopped.

I'm still getting console logs output, but nothing in structured logs or traces. Even running `azd up` from an older commit (that I believe used to work) doesn't fix the issue. Maybe a configuration problem in one of the containers?

Maybe I have to delete the entire resource group and re-publish from a new profile? But any suggestions to diagnose or fix when this happens to an existing Aspire application would be good.

**Comment by kevcrooks** [comment]
Just to confirm this, I created a new `azd` environment, and ran `azd up` - this created a new resource group in Azure with the dashboard, and the dashboard works perfectly fine.

However if I run `azd up` to the previous environment, the dashboard does _not_ show the structured logs or traces.

So I think either the Aspire dashboard is being corrupted over time, or there ends up being a problem inside the resource group that cannot be reverted.

@mitchdenny  any thoughts on this? I have both environments deployed so I can get more details, if you can let me know what to check?

**Comment by efonsecab** [comment]
I tried to reproduce it in a new app, no luck yet, however, comparing the Container Apps Environment logs I do see this, not sure if it is useful as a hint.

"ScaledObject doesn\u0027t have correct triggers specification"

This message appears only for the "Worker Service" projects

**Comment by mitchdenny** [comment]
@kvenkatrajan this might be an issue to raise the the ACA team. It seems like the problem is going to either be in the dashboard or in the resource server that ACA provides for the ACA-deployed dashboard to talk to.

**Comment by MrDeej** [comment]
Earlier I had logs in my test and dev environment. After I recreated the resource groups from scratch. Just had issue with production, but now all environments are without logs. I have not done anything, and the dev environment is just a clean Aspire starter kit solution with nothing. 

![image](https://github.com/user-attachments/assets/97cfe66d-c5e0-4ea5-a34b-98de421f448e)

I have a lot of logs in console
![image](https://github.com/user-attachments/assets/dc1eda68-ee83-4a6b-9692-aad3a050a027)

And the version differs from localhost:
![image](https://github.com/user-attachments/assets/e7a9ff03-f691-4b67-ad9d-68ad7ee2db1a)
![image](https://github.com/user-attachments/assets/966a2ea3-b034-4570-94c0-92fa45870bc1)


Any tips on how to debug would be greatly appreciated.

**Comment by kvenkatrajan** [comment]
@mitchdenny  sent it to ACA dev for investigation.

**Comment by snehapar9** [comment]
@MrDeej thanks for bringing this to our attention. This is a known issue, and we are working on its long-term fix. You were not seeing OTLP data because there was a mismatch in the OTLP Api key due to a race condition.
Can you please confirm you're seeing structured logs/metrics/traces now?

**Reply by efonsecab** [reply]
@snehapar9 is there any existent GitHub issue for this one?
Not sure about @MrDeej , but I do keep having the issue, tried re-deploying did not work, should I get latest daily build?

**Comment by MrDeej** [comment]
Hi @snehapar9 , thanks for response. No, I do not have any logs/metric/traces in any of my environments.

**Comment by efonsecab** [comment]
Today, out of nowhere, the structured logs started working, more or less, they began showing up for some of the projects in my solution, yet no all.
![image](https://github.com/user-attachments/assets/b1bb336f-6acb-4962-b575-253d96581080)

Resources
![image](https://github.com/user-attachments/assets/b55c63e4-0521-4ea7-9d21-80811ead9bda)


I did not change anything regarding configuration I just modified some existent pages and redeploy with azd up.

**Reply by snehapar9** [reply]
@efonsecab do all the apps deployed to your environment show up in the drop down? Do all these apps emit structured logs?

**Reply by efonsecab** [reply]
> @efonsecab do all the apps deployed to your environment show up in the drop down? Do all these apps emit structured logs?

They used too a couple of months ago, but at some point they all stopped showing up, and it was until a couple of days ago that some of them started to show up, but not all.

**Reply by snehapar9** [reply]
@efonsecab can you check if traces/structured logs show up if you hit the app endpoint?

**Reply by efonsecab** [reply]
> @efonsecab can you check if traces/structured logs show up if you hit the app endpoint?

The don't, it looks like this, I'm expecting the one named fairplaytube to show up, but it does not.
![image](https://github.com/user-attachments/assets/8c1966e9-a899-4b7a-838a-2da7902aadf2)


![image](https://github.com/user-attachments/assets/4b67adc7-478b-4fef-9c63-b12286311214)

**Comment by MrDeej** [comment]
My test environment dashboard is working again after re-deploy, but production did not work after re-deploy and still says "No structured logs found"

**Comment by MrDeej** [comment]
Hello. Any news here? I am still missing my logs in my production environment.

**Reply by mitchdenny** [reply]
Had a chat with our Dashboard folks and partners in ACA and it looks like a fix is already rolling out but hasn't reached your region yet.

**Comment by snehapar9** [comment]
@MrDeej the fix is being rolled out now. Has not been deployed to Norway East yet. 
It should be deployed in all regions by the end of this week.

**Comment by efonsecab** [comment]
Mine stopped showing up once again out of nowhere, apps are published in East US 2 Azure Region.
![image](https://github.com/user-attachments/assets/6970f59a-f2c2-49eb-a185-f0393e3aab74)

**Comment by MrDeej** [comment]
Today, all my dashboards are showing this one:
![image](https://github.com/user-attachments/assets/fb55ee82-d4ce-472e-8f7f-81b12caf7e02)

**Comment by efonsecab** [comment]
I can once again see everything in the Dashboard for the published apps.
![image](https://github.com/user-attachments/assets/bc8ae556-97bf-449d-8c04-5d59290796ec)

![image](https://github.com/user-attachments/assets/cfcc27bc-203a-41b3-adee-4a38d5de7da0)

**Comment by snehapar9** [comment]
@MrDeej Are you still seeing the 404 error?

**Reply by MrDeej** [reply]
Yes.

**Reply by snehapar9** [reply]
@MrDeej your container app may be scaled to 0. Can you please let me know if min replicas and max replicas are greater than 0?
Does the container app exist still?

**Reply by MrDeej** [reply]
I dont know wich container represents the dashboard, and I have 3 in my resource group. 

These 3 containers look like this:
![image](https://github.com/user-attachments/assets/bd5e6f4a-1c33-4cd8-b10f-4e602ed76993)

**Reply by snehapar9** [reply]
@MrDeej apologies for the delayed response. I looked at your dev, test and prod environments. The Aspire feature was disabled, and the underlying resources were deleted. Can you please re-enable it from the portal (in your container app environment page) and let me know if you can access the dashboard?

![image](https://github.com/user-attachments/assets/674f7b46-1fd7-4328-a758-40f921fe63d8)

**Comment by Sebosek** [comment]
Hi, I have the same problem but with the local environment. In my case, I run just the Aspire Dashboard in the container next to the OpenTelemetry Collector with my Hello World App.

My docker-compose file:

```yaml
version: '3.8'

services:
  boardgamesclub.backend:
    container_name: helloworld
    image: helloworld.backend
    build:
      context: ./backend/HelloWorld.Backend/HelloWorld.Backend
      dockerfile: Dockerfile
    environment:
      - ASPNETCORE_URLS=http://*:5000/
    ports:
      - "5000:5000"
    depends_on:
      - otel-collector
  
  otel-collector:
    container_name: otel-collector
    image: otel/opentelemetry-collector-contrib:latest
    command: [ "--config", "/etc/otel-collector-config.yaml" ]
    ports:
      - "4317:4317" # For gRPC
      - "8888:8888" # OpenTelemetry health check endpoint
    volumes:
      - ./backend/otel-collector-config.yml:/etc/otel-collector-config.yaml
  
  aspire:
    container_name: aspire
    image: mcr.microsoft.com/dotnet/aspire-dashboard:8.1.0
    ports:
      - "18888:18888" # For UI
      - "18889:18889"  # For OTLP protocol
    depends_on:
      - otel-collector
```

OpenTelemetry Conllector config
```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318

exporters:
  debug:
    verbosity: detailed

  otlp:
    endpoint: "aspire:18889"
    tls:
      insecure: true

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [otlp]
    metrics:
      receivers: [otlp]
      exporters: [otlp]
    logs:
      receivers: [otlp]
      exporters: [otlp]
```

C# instrumentation
```csharp
    public static IHostApplicationBuilder AddShared(this WebApplicationBuilder builder, IConfiguration configuration)
    {
        var serilog = new LoggerConfiguration();
        serilog.ReadFrom.Configuration(configuration);

        builder.Services.AddEndpointsApiExplorer();
        builder.Services.AddSwaggerGen();
        builder.Services.AddSerilog(serilog.CreateLogger());
        builder.Services.AddOptions<CorsConfig>()
            .Bind(configuration.GetSection(CorsConfig.KEY))
            .ValidateDataAnnotations()
            .ValidateOnStart();
        builder.Services.AddOptions<TelemetryConfig>()
            .Bind(configuration.GetSection(TelemetryConfig.KEY))
            .ValidateDataAnnotations()
            .ValidateOnStart();

        builder.Services.AddCors(setup =>
            setup.AddDefaultPolicy(policy =>
                policy.AddHostConfiguration(configuration)));

        var telemetry = configuration.GetSection(TelemetryConfig.KEY).Get<TelemetryConfig>()!;
        var serviceName = configuration.GetValue("ServiceName", defaultValue: "HelloWorld")!;
        var serviceVersion = typeof(Program).Assembly.GetName().Version?.ToString() ?? "dev";
        var serviceInstanceId = Environment.MachineName;

        builder.Logging
            .ClearProviders()
            .AddSerilog()
            .AddOpenTelemetry(logger =>
            {
                logger.SetResourceBuilder(ResourceBuilder.CreateDefault().AddService(
                    serviceName: serviceName,
                    serviceVersion: serviceVersion,
                    serviceInstanceId: serviceInstanceId));
                logger.IncludeScopes = true;
                logger.AddOtlpExporter(options =>
                {
                    options.Protocol = telemetry.Logging.Protocol;
                    options.Endpoint = telemetry.Logging.Endpoint;
                });
            });

        builder.Services
            .AddOpenTelemetry()
            .ConfigureResource(resource =>
                resource.AddService(
                    serviceName: serviceName,
                    serviceVersion: serviceVersion,
                    serviceInstanceId: serviceInstanceId))
            .WithTracing(tracing =>
            {
                tracing.AddAspNetCoreInstrumentation();
                tracing.AddHttpClientInstrumentation();
                tracing.AddSqlClientInstrumentation();
                tracing.AddOtlpExporter(options =>
                {
                    options.Protocol = telemetry.Tracing.Protocol;
                    options.Endpoint = telemetry.Tracing.Endpoint;
                });
            })
            .WithMetrics(metrics =>
            {
                metrics
                    .AddAspNetCoreInstrumentation()
                    .AddHttpClientInstrumentation()
                    .AddMeter("Microsoft.AspNetCore.Hosting")
                    .AddMeter("Microsoft.AspNetCore.Server.Kestrel")
                    .AddOtlpExporter(options =>
                    {
                        options.Protocol = telemetry.Metrics.Protocol;
                        options.Endpoint = telemetry.Metrics.Endpoint;
                    });
            });

        return builder;
    }
```

**Reply by davidfowl** [reply]
I don't think the problems are similar. The issues with the dashboard in azure containers apps are likely not the same as the ones you run into while running it locally.

**Comment by inlineHamed** [comment]
I have the same issue: on my local Windows machine, structured logs and metrics are available, but on my Debian 12 server, I have neither. I also tested the aspire-starter app on the server, but it does not show structured logs. Additionally, I tested it on my local WSL Debian environment, and there were no structured logs there either. so I think the issue is related to Debian/Linux.

version: 9.2.1

**Reply by davidfowl** [reply]
Did you setup https?

**Reply by inlineHamed** [reply]
That was the issue for me. thanks 🙏👍
it resolved by `dotnet dev-certs https --trust`

**Comment by VisualBean** [comment]
I came here from not having structured logs in the basic setup, until i noticed that 
Serilog.Sinks.OpenTelemetry weren't added to my ServiceDefaults.
```csharp
builder.Services.AddSerilog((services, loggerConfiguration) => loggerConfiguration
            .ReadFrom.Configuration(builder.Configuration)
            .Enrich.FromLogContext()
            .WriteTo.OpenTelemetry() // this did the trick for me
            .WriteTo.Console());
```
